lib/charms/opensearch/v0/helper_security.py

```diff
@@ -117,3 +123,213 @@ def to_pkcs8(private_key: str, password: Optional[str] = None) -> str:
+        alias = alias.split("friendlyName:")[-1]
+
+        certs[alias] = f"{start_cert_marker}{cert.split(start_cert_marker)[1]}{end_cert_marker}"
+
+    return certs
+
+
+def read_ca(alias: str, store_pwd: str, store_path: str) -> Optional[str]:
+    """Load stored CA cert."""
+    return (list_cas(store_pwd, store_path) or {}).get(alias)
+
+
+def remove_ca(alias: str, store_pwd: str, store_path: str) -> None:
+    """Remove old CA cert from trust store."""
+    if not exists(store_path):
+        return
+
+    keytool = "opensearch.keytool"
+
+    list_cmd = f"{keytool} -list -keystore {store_path} -alias {alias} -storetype PKCS12"
+    list_args = f"-storepass {store_pwd}"
+    try:
+        run_cmd(list_cmd, list_args)
+    except OpenSearchCmdError as e:
+        # This message means there was no "ca" alias or store before, if it happens ignore
+        if f"Alias <{alias}> does not exist" in e.out:
+            return
+
+    del_cmd = f"{keytool} -delete -keystore {store_path} -alias {alias} -storetype PKCS12"
+    del_args = f"-storepass {store_pwd}"
+    run_cmd(del_cmd, del_args)
+    logger.info(f"Removed {alias} from truststore.")
+
+
+def store_key_pair(
+    name: str, store_pwd: str, store_path: str, cert: str, key: str, key_pwd: str | None
+) -> None:
+    """Store cert in keystore."""
+    try:
+        os.remove(store_path)
+    except OSError:
+        pass
+
+    tmp_key = tempfile.NamedTemporaryFile(
+        mode="w+t", suffix=".pem", dir=os.path.dirname(store_path)
+    )
+    tmp_key.write(key)
+    tmp_key.flush()
+    tmp_key.seek(0)
+
+    tmp_cert = tempfile.NamedTemporaryFile(
+        mode="w+t", suffix=".cert", dir=os.path.dirname(store_path)
+    )
+    tmp_cert.write(cert)
+    tmp_cert.flush()
+    tmp_cert.seek(0)
+
+    cmd = f"openssl pkcs12 -export -in {tmp_cert.name} -inkey {tmp_key.name} -out {store_path} -name {name}"
+    args = f"-passout pass:{store_pwd}"
+    if key_pwd:
+        args = f"{args} -passin pass:{key_pwd}"
+
+    try:
+        run_cmd(cmd, args)
+        run_cmd(f"sudo chmod +r {store_path}")
+    except OpenSearchCmdError as e:
+        logging.error(f"Error storing the TLS certificates for {name}: {e}")
+    finally:
+        tmp_key.close()
+        tmp_cert.close()
+        logger.info(f"TLS certificate for {name} stored.")
+
+
+def get_cert_issuer(cert: str) -> Optional[str]:
+    """Retrieve the certificate issuer from a string certificate."""
+    # to make sure the content is processed correctly by openssl, temporary store it in a file
+    tmp_ca_file = tempfile.NamedTemporaryFile(mode="w+t", dir="/tmp")
+    tmp_ca_file.write(cert)
+    tmp_ca_file.flush()
+    tmp_ca_file.seek(0)
+
+    try:
+        return run_cmd(f"openssl x509 -in {tmp_ca_file.name} -noout -issuer").out
+    except OpenSearchCmdError as e:
+        logger.error(f"Error reading the current truststore: {e}")
+        return None
+    finally:
+        tmp_ca_file.close()
```

Comment:
This is also good but we can prefer to use context manager for guaranteed cleanup and file closing.
```suggestion
    with tempfile.NamedTemporaryFile(mode="w+t", dir="/tmp", delete=False) as f:
        f.write(cert)
        f.flush()
        path = f.name
    try:
        return run_cmd(f"openssl x509 -in {path} -noout -issuer").out
    except OpenSearchCmdError as e:
        logger.error(f"Error reading the current truststore: {e}")
        return None
    finally:
        os.remove(path)
```

Reply:
same for `store_key_pair`

Reply:
In this particular occurrence I believe the addition of a context manager is not very helpful, as the deletion is still warranted in the `finally`.
In addition to this, in the `store_key_pair` 2 context managers are needed, therefore 2 `os.remove` calls needed in the `finally`. Which ultimately would just add 1 level of indentation to the code